Fix Followers block error when social graph is hidden

obenland opened the pull request:
## Summary

Fixes the Fediverse Followers block throwing an error when the "Hide Social Graph" privacy option is enabled. This regression was introduced in #2294.

The issue occurs because the ActivityPub followers endpoint does not return `orderedItems` when the social graph is hidden, causing `Cannot read properties of undefined (reading 'length')` in the block editor.

This fix replaces the custom ActivityPub endpoint with WordPress's built-in REST API for fetching followers in the block editor, and properly respects the social graph visibility setting throughout.

## Changes

- Add REST query parameter support for `ap_actor` post type to filter by followed user.
- Expose follower meta fields (`_activitypub_acct`, `_activitypub_avatar_url`) to REST API.
- Refactor `edit.js` to use `useEntityRecords` instead of `apiFetch`.
- Filter users with hidden social graph from the block's user selector dropdown.
- Show warning notice in editor for existing blocks whose selected user has hidden their social graph.
- Add privacy check in `render.php` to skip rendering when social graph is hidden.
- REST API returns no results when querying followers for users with hidden social graph.
- Add `show_in_rest` to user meta registration for JavaScript access.
- Update unit tests to verify social graph visibility is respected.

## Test Plan

- [ ] Enable "Hide Social Graph" in user profile settings (under ActivityPub section).
- [ ] Insert Fediverse Followers block in a post.
- [ ] Verify the user with hidden social graph is not available in the user selector dropdown.
- [ ] If editing an existing block with a user who now has hidden social graph, verify warning notice appears.
- [ ] Verify the block does not display followers on the frontend when social graph is hidden.
- [ ] Disable "Hide Social Graph" setting.
- [ ] Verify block works correctly and user appears in dropdown again.
- [ ] Test with blog user (ID 0) - same behavior should apply.

<!-- start changelog -->
- [x] Automatically create a changelog entry from the details below.

<details open>
<summary>Changelog Entry Details</summary>

Significance:
- [ ] Patch: backwards-compatible bug fix.
- [x] Minor: added functionality in a backwards-compatible manner.
- [ ] Major: broke backwards compatibility; could be a minor in 0.x versions.

Type:
- [ ] Added: new feature.
- [x] Changed: updated existing functionality.
- [ ] Deprecated: marked for future removal.
- [ ] Removed: deleted code or feature.
- [ ] Fixed: bug fix.
- [ ] Security: fixed vulnerability.

Changelog entry: Respect "Hide Social Graph" setting in Followers block editor and frontend.

</details>
<!-- end changelog -->

obenland committed:
Fix Followers block error when social graph is hidden

The Followers block was throwing an error when the "hide social graph"
user setting was enabled because the ActivityPub endpoint does not return
orderedItems in that case.

This fix replaces the custom ActivityPub endpoint with WordPress's built-in
REST API for fetching followers in the block editor view. The block now
works reliably regardless of the social graph visibility setting.

Changes:
- Add REST query parameter support for ap_actor post type to filter by
  followed user
- Expose follower meta fields (acct, avatar_url) to REST API
- Refactor edit.js to use useEntityRecords instead of apiFetch
- Add unit tests for REST API filtering with visible/hidden social graph

obenland committed:
Add changelog entry

obenland committed:
Address PR feedback: add privacy check, sanitization, and title support

obenland commented on includes/class-post-types.php:
```diff
@@ -572,6 +597,45 @@ public static function register_ap_actor_rest_field() {
+					'key'   => Followers::FOLLOWER_META_KEY,
+					'value' => $following,
+				),
+			);
```
Fixed in [0460fe7](https://github.com/Automattic/wordpress-activitypub/pull/2625/commits/0460fe7170788fc3cef874c095170e6317c4868b).

obenland commented on includes/class-post-types.php:
```diff
@@ -572,6 +597,45 @@ public static function register_ap_actor_rest_field() {
+				array(
+					'key'   => Followers::FOLLOWER_META_KEY,
+					'value' => $following,
+				),
```
Fixed in [0460fe7](https://github.com/Automattic/wordpress-activitypub/pull/2625/commits/0460fe7170788fc3cef874c095170e6317c4868b).

obenland commented on includes/class-post-types.php:
```diff
@@ -572,6 +597,45 @@ public static function register_ap_actor_rest_field() {
+	public static function register_ap_actor_rest_query_params( $params ) {
+		$params['activitypub_following'] = array(
+			'description' => \__( 'Filter actors by who they follow (WordPress user ID).', 'activitypub' ),
+			'type'        => 'integer',
```
Fixed in [0460fe7](https://github.com/Automattic/wordpress-activitypub/pull/2625/commits/0460fe7170788fc3cef874c095170e6317c4868b).

pfefferle commented:
Several things:

* Shouldn't we disable the block completely when the feature is disabled? I think "Hide Followers and Following on Profile" includes the Blocks, not only the API-Endpoint.
* If you change the User in the Block Settings to the Blog-User, it keeps the Followers of the Former (real) User and adds some Dummy Users (Bob Developer, Eve Hacker, ...) and it does not store the change, because after a reload it falls back to the User.

More feedback inline.

pfefferle commented on includes/class-post-types.php:
```diff
@@ -572,6 +598,48 @@ public static function register_ap_actor_rest_field() {
+			'description'       => \__( 'Filter actors by who they follow (WordPress user ID).', 'activitypub' ),
+			'type'              => 'integer',
+			'sanitize_callback' => 'absint',
+			'validate_callback' => 'is_user_member_of_blog',
```
I am not sure the `is_user_member_of_blog` works here, because the blog user has ID `0`.

obenland replied:
Fixed in [5660fc9](https://github.com/Automattic/wordpress-activitypub/pull/2625/commits/5660fc9689fae6c673bc071692963ad31fffdf78).

pfefferle commented on includes/class-post-types.php:
```diff
@@ -572,6 +598,48 @@ public static function register_ap_actor_rest_field() {
+		$following = $request->get_param( 'activitypub_following' );
+
+		// Only filter if user ID provided and social graph is visible or current user can edit that user.
+		if ( $following && ( Actors::show_social_graph( $following ) || \current_user_can( 'edit_user', $following ) ) ) {
```
This check `\current_user_can( 'edit_user', $following )` might also not work for the Blog-User (id `0`).

obenland replied:
Fixed in [5660fc9](https://github.com/Automattic/wordpress-activitypub/pull/2625/commits/5660fc9689fae6c673bc071692963ad31fffdf78).

obenland commented:
> Shouldn't we disable the block completely when the feature is disabled? I think "Hide Followers and Following on Profile" includes the Blocks, not only the API-Endpoint.

I wondered the same and came to the conclusion that it's a block that needs to be inserted manually.

To complicate matters, I noticed this while testing #2617 and the Followers blocks I had in the test posts showed up broken. So they were already inserted and then I checked the box to hide followers. In that scenario, we'd probably want to not show them?

Nice catch with the blog user, obviously the one thing I didn't test. I tried to create e2e around this but was unsuccessful.

pfefferle committed:
Merge branch 'trunk' into fix/followers-block-entity-types

obenland committed:
Hide followers from REST API when social graph is hidden

- Add privacy check in render.php to skip rendering when social graph hidden
- Filter users with hidden social graph from block editor dropdown
- Show warning notice for existing blocks with hidden social graph users
- Fix REST API to return no results when social graph hidden (including blog user)
- Add show_in_rest to user meta registration for JS access
- Update test to verify social graph visibility is respected

obenland commented:
@pfefferle I updated the PR to 
* show a notice in the block if the selected user has sharing disabled.
* For newly inserted blocks it filters out all users (incl blog) that have it disabled.
* In the frontend it shows an HTML comment instead of the block when sharing is disabled.

obenland committed:
Merge branch 'trunk' into fix/followers-block-entity-types

pfefferle commented:
The idea behind hiding followers is that no one should be able to query the list at all. Re-enabling the AP-Actors Post-Type API endpoint makes this possible again through a different path. At that point, follower “hiding” relies on the assumption that no one will query the `wp/v2` endpoint.

pfefferle commented:
I think we should stick to the ActivityPub API endpoint and simply have the check if the user has the endpoint enabled or not!?

pfefferle committed:
Merge branch 'trunk' into fix/followers-block-entity-types

pfefferle committed:
Merge branch 'trunk' into fix/followers-block-entity-types

pfefferle committed:
Revert to ActivityPub endpoint, detect hidden social graph from API response.

- Revert REST API changes that exposed followers via /wp/v2/ap_actor
- Use apiFetch with ActivityPub endpoint instead of useEntityRecords
- Detect hidden social graph from API response (missing orderedItems)
- Show warning notice when social graph is hidden
- Keep user filtering in dropdown and frontend render check
- Add show_in_rest for blog's activitypub_hide_social_graph setting

obenland committed:
Use createInterpolateElement for hidden social graph notice.

Improves i18n by embedding the link within the translatable string
rather than concatenating separate strings.

pfefferle committed:
Merge branch 'trunk' into fix/followers-block-entity-types

obenland committed:
Use canUser selector for capability checking in Followers block

- Replace currentUser.capabilities?.manage_options with canUser selector
- Check manage_options via canUser('update', { kind: 'root', name: 'site' })
- Move authorId useSelect earlier to fix reference issues
- Handle hidden social graph notice for both blog and inherit modes
- Auto-switch to first available user only for explicit user selections
- Remove debug console.log statement
